Feature/task details screen

skaik-mo opened the pull request:
This PR introduces the `TaskDetailsScreen` composable along with its associated UI state and components.

- Created `TaskDetailsScreen.kt` to display detailed information about a task.
- Implemented `TaskDetailsUiState.kt` to manage the state of the task details screen.
- Developed `TaskStatusUiState.kt` to represent different task statuses (TODO, IN_PROGRESS, DONE) with corresponding string resources and colors.
- Added new components for the task details screen:
    - `TaskActionButtons.kt`: Provides action buttons for task management.
    - `TaskDetailsDivider.kt`: A visual separator for different sections.
    - `TaskStatusAndPrioritySection.kt`: Displays the task's status and priority.
    - `TaskStatusBadge.kt`: A badge to visually indicate the task's status.
- Added a new string resource for "Task details" and "Move to".
- Moved `TaskStatusUiState` enum from `TaskUiState.kt` to its own file `TaskStatusUiState.kt`.
- Updated `CategoryUiState.kt` to provide a default image resource.

skaik-mo committed:
feat: implement TaskDetailsScreen

This commit introduces the `TaskDetailsScreen` composable along with its associated UI state and components.

- Created `TaskDetailsScreen.kt` to display detailed information about a task.
- Implemented `TaskDetailsUiState.kt` to manage the state of the task details screen.
- Developed `TaskStatusUiState.kt` to represent different task statuses (TODO, IN_PROGRESS, DONE) with corresponding string resources and colors.
- Added new components for the task details screen:
    - `TaskActionButtons.kt`: Provides action buttons for task management.
    - `TaskDetailsDivider.kt`: A visual separator for different sections.
    - `TaskStatusAndPrioritySection.kt`: Displays the task's status and priority.
    - `TaskStatusBadge.kt`: A badge to visually indicate the task's status.
- Added a new string resource for "Task details".
- Moved `TaskStatusUiState` enum from `TaskUiState.kt` to its own file `TaskStatusUiState.kt`.
- Updated `CategoryUiState.kt` to provide a default image resource.

skaik-mo committed:
Merge branch 'feature/tudee-bottm-sheet-component' into feature/task-details-screen

* feature/tudee-bottm-sheet-component:
  add Tudee Bottom Sheet

skaik-mo committed:
Add TaskDetailsViewModel and integrate with TaskDetailsScreen

This commit introduces a `TaskDetailsViewModel` to manage the UI state for the task details screen.

The `TaskDetailsUiState` has been updated to include an `isVisible` property to control the visibility of the task details bottom sheet.

The `TaskDetailsScreen` now uses the `TaskDetailsViewModel` to manage its state and interactions. The content of the screen is wrapped in a `TudeeBottomSheet` which can be dismissed by calling the `hidden` function in the ViewModel.

The `uiModule` in DI has been updated to provide an instance of `TaskDetailsViewModel`.

skaik-mo committed:
Merge branch 'develop' into feature/task-details-screen

* develop:
  add Secondary Button
  add Primary Button
  add content to Tudee Button
  add content to Tudee Text Button

skaik-mo committed:
Implement TaskActionButtons and task status update logic

- Add TaskActionButtons composable with edit and move task status functionality.
- Implement `onEditTaskClick` and `onMoveTaskStatusClick` in `TaskDetailsViewModel`.
- Add `getNextState` function to `TaskStatusUiState` to determine the next task status.
- Add `getStatusText` function to `TaskStatusUiState` to get status as string.
- Update `TaskDetailsUiState` to include new properties for icons and button titles.
- Add pencil edit icon.
- Add string resource for "Move to".
- Animate background and text color changes in `TaskStatusBadge`.
- Show/hide action buttons based on task completion status with animation.

skaik-mo committed:
Merge branch 'develop' into feature/task-details-screen

* develop:
  update floating action button
  improve Tudee Bottom sheet by make it expanded and add the navigation Bars Padding

hazemka commented on app/src/main/java/com/qudus/tudee/ui/screen/task_details/TaskDetailsScreen.kt:
```diff
@@ -0,0 +1,114 @@
+                TaskDetailsDivider()
+            }
+            item {
+                TaskStatusAndPrioritySection(state)
```
if you pass what is needed from state this very good why ?
1- Encapsulation (Each composable should only know what it cares about, Keeps components reusable and easier to test).
2- Performance (fewer recompositions).
3- Readability (Passing individual properties makes the data flow clear).

hazemka reviewed (changes requested):
change that I mentioned in comments in TaskDetailsScreen.kt

hazemka commented on app/src/main/java/com/qudus/tudee/ui/screen/task_details/TaskDetailsScreen.kt:
```diff
@@ -0,0 +1,114 @@
+            }
+            item {
+                TaskActionButtons(
+                    state = state,
```
here the same of previous comment

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/state/PriorityUiState.kt:
```diff
@@ -1,13 +1,36 @@
+        LOW -> R.string.low.toStringResource()
+        MEDIUM -> R.string.medium.toStringResource()
+        HIGH -> R.string.high.toStringResource()
+    }
```
Your `UiState `should be plain and **pure data**, not tied to Android framework elements like @`Composable`, `Painter`, [Color](url), or `string resources`.

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/state/TaskStatusUiState.kt:
```diff
@@ -0,0 +1,39 @@
+        TODO -> IN_PROGRESS
+        else -> DONE
+    }
+
```
same here,  Your UiState should be plain and pure data, not tied to Android framework elements like `@Composable, Painter, Color, or string resources`.

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/util/extension/IntExtension.kt:
```diff
@@ -0,0 +1,16 @@
+@Composable
+fun Int.toStringResource(): String {
+    return stringResource(this)
+}
```
What will happen here if I pass a general number instead of a resource ID ?
try to use a safer way to do that

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/state/CategoryUiState.kt:
```diff
@@ -1,11 +1,12 @@
     val title:String="",
      @DrawableRes
-     val image: Int? = null,
+     val image: Int = R.drawable.icon_book_open,
```
The UiState should hold only pure testable data

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/screen/task_details/components/TaskStatusBadge.kt:
```diff
@@ -0,0 +1,64 @@
+
+@Composable
+fun TaskStatusBadge(
+    taskStatusUiState: TaskStatusUiState,
```
To improve reusability, consider making this composable accept **pure, generic data** instead of a **specific** `TaskStatusUiState`. For example, you can pass a label, icon, and color directly. That way, this component becomes more flexible and can be reused across different features without depending on a single UI state's structure.

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/screen/task_details/components/TaskStatusAndPrioritySection.kt:
```diff
@@ -0,0 +1,28 @@
+import androidx.compose.ui.unit.dp
+import com.qudus.tudee.ui.designSystem.component.priority.PriorityBadge
+import com.qudus.tudee.ui.screen.task_details.TaskDetailsUiState
+
+@Composable
+fun TaskStatusAndPrioritySection(taskDetailsUiState: TaskDetailsUiState) {
+    Row(
+        horizontalArrangement = Arrangement.spacedBy(8.dp),
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        TaskStatusBadge(
+            taskStatusUiState = taskDetailsUiState.taskUiState.taskStatusUiState,
+            modifier = Modifier.height(28.dp)
+        )
+        PriorityBadge(
+            priority = taskDetailsUiState.taskUiState.taskPriority,
+            modifier = Modifier.height(28.dp)
+        )
+    }
+}
```
Since this section (`TaskStatusAndPrioritySection`) is tightly coupled to the  `Task Details screeen`, I’d suggest keeping it in the same file as the screen.

skaik-mo replied:
I separated it within the same screen package.

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/screen/task_details/components/TaskActionButtons.kt:
```diff
@@ -0,0 +1,72 @@
+            modifier = Modifier.padding(top = 24.dp)
+        ) {
+            TudeeButton(
+                isLoading = false,
```
Instead of hardcoding isLoading = false, it’s better to bind it to the screen’s state. This way, the button can react to the actual loading state (e.g., when a task is being updated), ensuring the UI stays consistent with the app’s logic.

skaik-mo replied:
I prefer not to include that on task details screen, as it contains two buttons, one for navigating to the task editor and another for changing the task state.

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/ui/screen/task_details/TaskDetailsUiState.kt:
```diff
@@ -0,0 +1,13 @@
+import com.qudus.tudee.ui.state.TaskUiState
+
+data class TaskDetailsUiState(
+    val taskUiState: TaskUiState = TaskUiState(),
+    val editIcon: Int = R.drawable.icon_pencil_edit,
+    val editIconDescription: String = "Pencil Edit",
+    var moveButtonTitle: Int = R.string.move_to,
+    var isVisible: Boolean = true,
+    var isTaskCompleted: Boolean = false
```
- These resources should ideally be resolved in the UI layer (composables part), not inside the state itself.
- 
- I noticed the `isLoading: Boolean` property is missing from the state. Adding it would help properly represent loading or processing states in the UI and improve state management during async operations.

skaik-mo replied:
I receive the task directly, so there's no need for a loading state.

skaik-mo committed:
Merge branch 'develop' into feature/task-details-screen

* develop:
  removed commented out code and redundant heightIn max value for CategoryBadgeItem in AddTaskScreen
  improve priority resources usage
  update AddTaskScreen buttons to use Text composable for labels
  remove logging from `AddTaskViewModel`
  remove redundant logging in `getCategories` method
  Replace FlowRow with adaptive LazyVerticalGrid in AddTaskScreen for category display
  update TudeeBottomSheet to prevent hiding on user interaction
  Trim title input before validation
  Implement add task screen with ViewModel and UI state
  This commit introduces new mapper files to convert between domain entities and UI state objects for tasks and categories. It also adds utility functions for retrieving localized strings and icon resources based on default category types. Additionally, functions for generating error messages for task title and category selection have been included.
  Create TitledSection, PriorityChip, and TudeeBottomSheet composables
  This update introduces the ability to display images from both resources and file paths within the `CategoryBadgeItem`. It also modifies the click handler to pass the item's ID.
  This commit introduces new drawable resources for a calendar "add" icon and a birthday cake icon.
  Refactor: update Category entity and introduce ValidationException
  Refactor Koin modules for services, data, and UI
  Refactor: implement Room callback to populate default categories and enhance data layer
  Add BaseViewModel helpers for suspend actions and flow collection
  Solve string resource error and wrong import calling

# Conflicts:
#	app/src/main/java/com/qudus/tudee/app/di/uiModule.kt
#	app/src/main/res/values/strings.xml

skaik-mo committed:
Refactor: Update TaskDetailsScreen and related components

This commit refactors the TaskDetailsScreen and its associated components for improved clarity and maintainability.

Key changes:
- Extracted `CategoryIcon` into its own composable function within `CategoryBadgeItem.kt`.
- Updated `TaskActionButtons` to accept specific parameters for visibility, icons, and text, rather than relying on the entire `TaskDetailsUiState`.
- Modified `TaskStatusAndPrioritySection` to accept `TaskStatusUiState` and `PriorityUiState` directly instead of the full `TaskDetailsUiState`.
- Updated `TaskStatusBadge` to take `backgroundColor`, `textColor`, and `title` as parameters for better reusability and clarity.
- Updated `PriorityBadge` to use `Theme.textStyle.label.small` for consistency and rounded corners with 100.dp.
- Simplified state access in `PrioritySelector.kt`, `TaskStatusUiState.kt`, and `PriorityUiState.kt` by using extension functions for getting colors, icons, and labels.
- Updated `TaskDetailsScreen.kt` to pass the necessary individual state properties to its child components.

skaik-mo committed:
Refactor: Update TaskStatusUiState and related components

- Removed `@StringRes` from `TaskStatusUiState` enum and updated `getStatusText()` to use a `when` expression for string resource retrieval.
- Updated `TaskStatusBadge` and `TaskDetailsScreen` to use the modified `getStatusText()`.
- Refactored `PriorityBadge` to accept `backgroundColor`, `icon`, and `title` as parameters.
- Updated `TaskStatusAndPrioritySection` to pass necessary parameters to `PriorityBadge` and `TaskStatusBadge`.
- Removed logging from `TaskDetailsViewModel`.
- Added new string resources for "Task details" and "Move to" and organized existing ones.
- Corrected Arabic translation for "todo" status.

skaik-mo committed:
Merge branch 'develop' into feature/task-details-screen

* develop:
  update dimensions
  update dimensions
  resolve conflict
  update on boarding screen
  update on boarding screen
  update strings resource file
  update ui module
  refactor on boarding screen
  update on boarding ui state
  refactor on boarding screen
  refactor on boarding screen
  complete on boarding feature
  add resource file

skaik-mo committed:
Refactor: Implement task state transition logic

This commit introduces the `moveToState` functionality in `TaskService` and `TaskServiceImpl` to allow tasks to transition between states (TODO, IN_PROGRESS, DONE).

Key changes:
- Added `moveToState` method to `TaskService` and its implementation in `TaskServiceImpl`.
- Updated `TaskDetailsViewModel` to utilize `moveToState` when the "Move Task Status" button is clicked.
- `TaskActionButtons` now dynamically displays the next state in the "Move Task Status" button text.
- Introduced `StateMapper` to convert between `TaskStatusUiState` and domain `State`.
- Modified `TaskUiState` to use `Long` for `taskId`.
- Removed redundant properties from `TaskDetailsUiState` as they are now derived or handled elsewhere.

skaik-mo committed:
Merge branch 'develop' into feature/task-details-screen

* develop:
  refactor wrap service call
  delete files not required for splash screen api
  refine splash screen
  add wrap service call
  implement splash screen

Farah315 commented on app/src/main/java/com/qudus/tudee/ui/screen/task_details/components/TaskStatusBadge.kt:
```diff
@@ -0,0 +1,81 @@
+    val animatedColor by animateColorAsState(targetValue = textColor)
+    Row(
+        verticalAlignment = Alignment.CenterVertically,
+        horizontalArrangement = Arrangement.spacedBy(4.dp),
```
u  can use Dimension like spacing4

Farah315 reviewed (commented):
good work

skaik-mo committed:
Refactor: Enhance TaskDetailsScreen with data fetching and UI state management

This commit introduces data fetching capabilities to the `TaskDetailsScreen` by implementing `getTaskById` in `TaskServiceImpl` and `getCategoryById` in `CategoryServiceImpl`.

Key changes include:

- **Data Fetching:** Implemented logic in `TaskDetailsViewModel` to fetch task and category details.
- **UI State Management:**
    - Added mappers to convert domain entities (Task, Category, Priority, State) to their respective UI state representations (`TaskUiState`, `CategoryUiState`, `PriorityUiState`, `TaskStatusUiState`).
    - Introduced `isMoveOperationLoading` in `TaskDetailsUiState` to manage the loading state of the move task operation.
    - Updated `TaskDetailsScreen` to reflect loading states and display fetched data.
- **UI Enhancements:**
    - Added padding to `TaskStatusAndPrioritySection`.
    - `TaskActionButtons` now reflects the loading state of the move operation.
- **Mapper Additions:**
    - `data/mapper/TaskMapper.kt`: Added `toTask()` to convert `TaskDto` to `Task`.
    - `ui/mapper/CategoryMapper.kt`: Added `toCategoryUiState()` to convert `Category` to `CategoryUiState`.
    - `ui/mapper/PriorityMapper.kt`: New file with mappers between `Priority` and `PriorityUiState`.
    - `ui/mapper/StateMapper.kt`: Added `toTaskStatusUiState()` to convert `State` to `TaskStatusUiState`.
    - `ui/mapper/TaskMapper.kt`: Added `toTaskUiState()` to convert `Task` to `TaskUiState`.
- **State Updates:**
    - `CategoryUiState`: Added `defaultCategoryType` field.

skaik-mo committed:
Refactor: Update TaskDetailsScreen to handle exceptions and improve UI

This commit introduces several changes to the TaskDetailsScreen and related components:

- **Exception Handling:**
    - The `TaskDetailsUiState` now includes an `exception` field to store `TudeeExecption`.
    - `TaskDetailsViewModel` now updates the `exception` field in the UI state when errors occur during data fetching or task operations.
    - A new composable `DataErrorContent` is introduced to display error messages within the `TaskDetailsScreen` when an exception is present in the UI state.
- **UI Improvements:**
    - `TudeeBottomSheet` now uses `statusBarsPadding` instead of `navigationBarsPadding` for better layout consistency.
    - `TaskDetailsScreen` now correctly displays category icons for both predefined and custom categories by using `rememberAsyncImagePainter` for file paths.
- **Minor Fixes:**
    - The `taskId` in `TaskDetailsViewModel` has been updated (though this is likely a placeholder).
    - `CategoryUiState`'s `image` field is now a `String` to accommodate file paths, removing the `@DrawableRes` annotation.
    - Removed an unused import of `AddTaskScreen` from `MainActivity`.

skaik-mo committed:
Refactor: Update TaskDetailsScreen to use Theme.dimension for spacing

This commit refactors the `TaskDetailsScreen` and its related components to use spacing values from `Theme.dimension` instead of hardcoded Dp values.

This includes updates to:
- `DataErrorContent.kt`
- `TaskDetailsScreen.kt`
- `TaskStatusBadge.kt`
- `TaskStatusAndPrioritySection.kt`
- `TaskActionButtons.kt`
- `TaskDetailsDivider.kt`

Additionally, `spacing6` has been added to `Dimension.kt`.

mohammed-akkad commented on app/src/main/java/com/qudus/tudee/ui/screen/task_details/TaskDetailsUiState.kt:
```diff
@@ -0,0 +1,14 @@
+data class TaskDetailsUiState(
+    val taskUiState: TaskUiState = TaskUiState(),
+    var categoryUiState: CategoryUiState = CategoryUiState(),
+    var isVisible: Boolean = true,
```
Could you explain why var was chosen over val here?

skaik-mo replied:
I thought I was replacing the value, but I ended up using copy instead.

skaik-mo committed:
Refactor: Make `TaskDetailsUiState` properties immutable

The properties `categoryUiState`, `isVisible`, `isTaskCompleted`, `isMoveOperationLoading`, and `exception` in the `TaskDetailsUiState` data class have been changed from mutable (`var`) to immutable (`val`).

skaik-mo committed:
Merge branch 'develop' into feature/task-details-screen
